File: gui/old-modules/gator.rb
name 'GATOR'
hp 9

hue = 240
foreground [hue, 100, 30]
background [hue, 100, 97]

top = hp2mm(3.5)
dy = hp2mm(1.75)

left = hp2mm(1.5)
lc = hp2mm(3.25)
rc = hp2mm(5.75)
right = hp2mm(7.5)

(1..8).each do |r|
  row = top + (r - 1) * dy
  port(x: left, y: row)
  button(x: lc, y: row)
  button(x: rc, y: row)
  port(x: right, y: row)
end

row = top + 8.75 * dy
output_port(x: left, y: row, label: 'AND')
output_port(x: lc, y: row, label: 'NAND')
output_port(x: rc, y: row, label: 'OR')
output_port(x: right, y: row, label: 'NOR')

row += hp2mm(2.75)
output_port(x: left, y: row, label: 'EVEN')
output_port(x: lc, y: row, label: 'ODD')
output_port(x: rc, y: row, label: 'XOR')
output_port(x: right, y: row, label: 'XNOR')

File: panelgen/internal/panel/gator.go
package panel

import (
	"dhemery.com/DHE-Modules/panelgen/internal/control"
	"dhemery.com/DHE-Modules/panelgen/internal/svg"
)

func init() {
	registerBuilder("gator", gator)
}

func gator() *Panel {
	const (
		hue = 220.0
		hp  = 9
	)
	var (
		fg = svg.HslColor(hue, 1, .3)
		bg = svg.HslColor(hue, 1, .97)
	)
	p := NewPanel("GATOR", hp, fg, bg, "gator")

	const (
		top         = 3.5 * mmPerHp
		dy          = 1.75 * mmPerHp
		left        = 1.5 * mmPerHp
		leftCenter  = 3.25 * mmPerHp
		width       = float64(hp) * mmPerHp
		rightCenter = width - leftCenter
		right       = width - left
	)
	var (
		port   = control.Port(fg, bg)
		button = control.Button(bg, fg)
	)
	for row := 0; row < 8; row++ {
		y := top + float64(row)*dy
		p.Install(left, y, port)
		p.Install(leftCenter, y, button)
		p.Install(rightCenter, y, button)
		p.Install(right, y, port)
	}

	y := top + 8.75*dy
	p.OutPort(left, y, "AND")
	p.OutPort(leftCenter, y, "NAND")
	p.OutPort(rightCenter, y, "OR")
	p.OutPort(right, y, "NOR")

	y += 2.75 * mmPerHp
	p.OutPort(left, y, "EVEN")
	p.OutPort(leftCenter, y, "ODD")
	p.OutPort(rightCenter, y, "XOR")
	p.OutPort(right, y, "XNOR")

	return p
}
